core/src/main/java/com/gentics/mesh/plugin/manager/MeshPluginManagerImpl.java

```diff
@@ -79,30 +90,92 @@
 
 	public static final String PLUGINS_DIR_CONFIG_PROPERTY_NAME = "pf4j.pluginsConfigDir";
 
+	private final Lazy<Vertx> vertx;
+
 	private final PluginFactory pluginFactory;
 
 	private final MeshOptions options;
 
+	private final List<PluginWrapper> restartingPlugins = Collections.synchronizedList(new ArrayList<>(0));
+
+	private final List<PluginWrapper> startingPlugins = Collections.synchronizedList(new ArrayList<>(1));
+
 	// We track our own plugin status since the PF4J state is not extendible.
 	private final Map<String, PluginStatus> pluginStatusMap = new HashedMap<>();
 
 	private final DelegatingPluginRegistry pluginRegistry;
 
 	@Inject
-	public MeshPluginManagerImpl(MeshOptions options, MeshPluginFactory pluginFactory, DelegatingPluginRegistry pluginRegistry) {
+	public MeshPluginManagerImpl(MeshOptions options, MeshPluginFactory pluginFactory,
+			DelegatingPluginRegistry pluginRegistry, Database database, Lazy<Vertx> vertx) {
 		this.pluginFactory = pluginFactory;
 		this.options = options;
 		this.pluginRegistry = pluginRegistry;
+		this.vertx = vertx;
 		delayedInitialize();
 	}
 
 	protected void delayedInitialize() {
 		super.initialize();
 	}
 
+	protected void registerEventListeners() {
+		if (options.getClusterOptions().isEnabled()) {
+			EventBus eb = vertx.get().eventBus();
+
+			eb.consumer(CLUSTER_DATABASE_CHANGE_STATUS.address, (Message<JsonObject> handler) -> {
+				JsonObject info = handler.body();
+				Boolean isMe = info.getBoolean("isMe");
+
+				if (isMe != null && isMe) {
+					DB_STATUS status = DB_STATUS.valueOf(info.getString("status"));
+					log.info("Received local database status update - " + status.name());
+
+					switch (status) {
+					case NOT_AVAILABLE:
+					case OFFLINE:
+					case SYNCHRONIZING:
+					case BACKUP:
+						handleDatabaseAvailability(false);
+						break;
+					case ONLINE:
+						handleDatabaseAvailability(true);
+						break;
+					}
+				}
+			});
+			log.info("Distributed DB status event listener registered.");
+		}
+	}
+
+	protected synchronized void handleDatabaseAvailability(boolean available) {
+		if (available) {
+			startPluginsFrom(restartingPlugins);
+		} else {
+			synchronized (startingPlugins) {
```

Comment:
This `synchronized` ensures that we don't mess with non-ONLINE state changes and the initial plugin initialization. Those should never be called from each other, to avoid deadlocks.